tests/requestparsing/contentlengthandtransferencoding_test.go

```diff
@@ -0,0 +1,446 @@
+		"\r\n" +
+		"5\r\n" +
+		"ABCDE\r\n" +
+		"0\r\n" +
+		"\r\n")
+
+	resp, err := requesttesting.MakeRequest(context.Background(), request, func(r *http.Request) {
+		want := map[string][]string{"Content-Length": []string{"11"}}
+		if diff := cmp.Diff(want, map[string][]string(r.Header)); diff != "" {
+			t.Errorf("r.Header mismatch (-want +got):\n%s", diff)
+		}
+
+		if diff := cmp.Diff([]string(nil), r.TransferEncoding); diff != "" {
+			t.Errorf("r.TransferEncoding mismatch (-want +got):\n%s", diff)
+		}
+
+		if r.ContentLength != 11 {
+			t.Errorf("r.ContentLength got: %v want: 11", r.ContentLength)
+		}
+
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			t.Errorf("ioutil.ReadAll(r.Body) got: %v want: nil", err)
+		}
+
+		if got := string(body); got != "5\r\nABCDE\r\n0" {
+			t.Errorf(`r.Body got: %q want: "5\r\nABCDE\r\n0"`, got)
+		}
+	})
+	if err != nil {
+		t.Errorf("ERROR: %v", err)
+	}
+
+	if !bytes.HasPrefix(resp, []byte(statusOK)) {
+		got := string(resp[:bytes.IndexByte(resp, '\n')+1])
+		t.Errorf("status code got: %q want: %q", got, statusOK)
+	}
+}
+
+func TestTransferEncodingListChunkedFirst(t *testing.T) {
+	request := []byte("GET / HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" +
+		"Content-Length: 11\r\n" +
+		"Transfer-Encoding: chunked, xyz, asdf\r\n" +
+		"\r\n" +
+		"5\r\n" +
+		"ABCDE\r\n" +
+		"0\r\n" +
+		"\r\n")
+
+	resp, err := requesttesting.MakeRequest(context.Background(), request, nil)
+	if err != nil {
+		t.Errorf("ERROR: %v", err)
+	}
+
+	if !bytes.HasPrefix(resp, []byte(statusNotImplemented)) {
+		got := string(resp[:bytes.IndexByte(resp, '\n')+1])
+		t.Errorf("status code got: %q want: %q", got, statusNotImplemented)
+	}
+}
+
+func TestTransferEncodingListChunkedIdentity(t *testing.T) {
+	request := []byte("GET / HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" +
+		"Content-Length: 11\r\n" +
+		"Transfer-Encoding: chunked, identity\r\n" +
+		"\r\n" +
+		"5\r\n" +
+		"ABCDE\r\n" +
+		"0\r\n" +
+		"\r\n")
+
+	resp, err := requesttesting.MakeRequest(context.Background(), request, func(r *http.Request) {
+		want := map[string][]string{}
+		if diff := cmp.Diff(want, map[string][]string(r.Header)); diff != "" {
+			t.Errorf("r.Header mismatch (-want +got):\n%s", diff)
+		}
+
+		if diff := cmp.Diff([]string{"chunked"}, r.TransferEncoding); diff != "" {
+			t.Errorf("r.TransferEncoding mismatch (-want +got):\n%s", diff)
+		}
+
+		if r.ContentLength != -1 {
+			t.Errorf("r.ContentLength got: %v want: -1", r.ContentLength)
+		}
+
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			t.Errorf("ioutil.ReadAll(r.Body) got: %v want: nil", err)
+		}
+
+		if got := string(body); got != "ABCDE" {
+			t.Errorf(`r.Body got: %q want: "ABCDE"`, got)
+		}
+	})
+	if err != nil {
+		t.Errorf("ERROR: %v", err)
+	}
+
+	if !bytes.HasPrefix(resp, []byte(statusOK)) {
+		got := string(resp[:bytes.IndexByte(resp, '\n')+1])
+		t.Errorf("status code got: %q want: %q", got, statusOK)
+	}
+}
```

Comment:
It seems to me all of this could be a tabletest with:
```
tests := []struct{
  name string
  request string
  verifyHandler func(*http.Request)
  wantStatus string
} {
  // ...
}
```

Right?

Reply:
Yes! But all `verifyHandler`s would look really similar.
You could also convert most of the tests in the file to a table test that looks like this:
```go
tests := []struct{
  name string
  request string
  want wantStruct{
    headers map[string][]string
    contentLength int
    transferEncoding []string
    body string
  }
} {
  // ...
}
```
And put all tests with non-`200 OK` status codes in a separate table test.
What do you think about this solution?

Reply:
I implemented my solution in 83d3413. What do you think?